lib/plugins/bull-mq-metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,164 @@
+import { Queue, QueueEvents } from 'bullmq'
+import type { Redis } from 'ioredis'
+import * as prometheus from 'prom-client'
+
+import type { BullMqMetricsPluginOptions } from '../bullMqMetricsPlugin'
+
+type Metrics = {
+  completedGauge: prometheus.Gauge<never>
+  activeGauge: prometheus.Gauge<never>
+  delayedGauge: prometheus.Gauge<never>
+  failedGauge: prometheus.Gauge<never>
+  waitingGauge: prometheus.Gauge<never>
+  completedDuration: prometheus.Histogram<never>
+  processedDuration: prometheus.Histogram<never>
+}
+
+type MetricCollectorOptions = Required<Omit<BullMqMetricsPluginOptions, 'redisClient'>>
+
+export class MetricsCollector {
+  private readonly options: MetricCollectorOptions
+  private readonly metrics: Metrics
+  private active = true
+
+  constructor(
+    private readonly redis: Redis,
+    private readonly registry: prometheus.Registry,
+    options: Partial<MetricCollectorOptions>,
+  ) {
+    this.options = {
+      bullMqPrefix: 'bull',
+      metricsPrefix: 'bullmq',
+      excludedQueues: [],
+      collectionIntervalInMs: 5000,
+      histogramBuckets: [20, 50, 150, 400, 1000, 3000, 8000, 22000, 60000, 150000],
+      ...options,
+    }
+
+    this.metrics = this.registerMetrics(this.registry, this.options)
+  }
+
+  async start() {
+    const queueNames = await this.findQueues(this.redis, this.options)
+    await Promise.all(
+      queueNames.map((name) => this.observeQueue(name, this.redis, this.metrics, this.options)),
+    )
+  }
+
+  stop() {
+    this.active = false
+  }
+
+  private registerMetrics(
+    registry: prometheus.Registry,
+    { metricsPrefix, histogramBuckets }: MetricCollectorOptions,
+  ): Metrics {
+    const metrics: Metrics = {
```

Comment:
I would recommend setting some kind of `metricsInitted` flag, because prometheus will break if you try to create same metric twice. And this can easily happen e. g. when you are creating test context or app several times without purging Prometheus registry first